From d470dedf740961ab9a50a04740f9a6909e1e0bfc Mon Sep 17 00:00:00 2001
From: Nick Wellnhofer <wellnhofer@aevum.de>
Date: Mon, 5 Aug 2024 14:58:37 +0200
Subject: [PATCH] parser: Fix error handling after reaching limit

Mark document as non-wellformed and stop parser even if error limit was
reached.

Regressed in abd74186.
---
 parserInternals.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/parserInternals.c b/parserInternals.c
index 4456896e8..1ab0b11c6 100644
--- a/parserInternals.c
+++ b/parserInternals.c
@@ -253,11 +253,11 @@ xmlCtxtVErr(xmlParserCtxtPtr ctxt, xmlNodePtr node, xmlErrorDomain domain,
 
     if (level == XML_ERR_WARNING) {
         if (ctxt->nbWarnings >= XML_MAX_ERRORS)
-            return;
+            goto done;
         ctxt->nbWarnings += 1;
     } else {
         if (ctxt->nbErrors >= XML_MAX_ERRORS)
-            return;
+            goto done;
         ctxt->nbErrors += 1;
     }
 
@@ -308,6 +308,7 @@ xmlCtxtVErr(xmlParserCtxtPtr ctxt, xmlNodePtr node, xmlErrorDomain domain,
         return;
     }
 
+done:
     if (level >= XML_ERR_ERROR)
         ctxt->errNo = code;
     if (level == XML_ERR_FATAL) {
-- 
GitLab

